File: leetcode/416.partition-equal-subset-sum.py
# @leet start
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        total = sum(nums)

        if total % 2 == 1:
            return False

        target = total // 2
        sums = {0}

        for num in nums:
            new_sums: set[int] = set()

            for sum_ in sums:
                new_sum = num + sum_

                if new_sum == target:
                    return True

                if new_sum < target:
                    new_sums.add(new_sum)

            sums.update(new_sums)

        return False


# @leet end

File: leetcode/416.partition-equal-subset-sum.js
// @leet start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function (nums) {
  const sum = nums.reduce((acc, x) => acc + x);

  if (sum % 2 !== 0) {
    return false;
  }

  const target = sum / 2;
  const possible = Array(target + 1).fill(false);
  possible[0] = true;

  for (const num of nums) {
    if (num <= target && possible[target - num]) {
      return true;
    }

    for (let i = target - 1; i >= num; i--) {
      possible[i] ||= possible[i - num];
    }
  }

  return possible[target];
};
// @leet end
